File: packages/hqtui/src/widgets/table.ts
import type { Surface, Align } from "../surface.ts";
import { Attr, type Style } from "../buffer.ts";
import type { Color } from "../color.ts";
import { mix } from "../color.ts";
import { fit, stringWidth, truncate } from "../unicode.ts";
import { solve } from "../layout.ts";
import { elevate } from "../theme.ts";

export interface Column<Row = Record<string, unknown>> {
  /** Property to read, or use `render` for computed cells. */
  key: string;
  title?: string;
  width?: number | string;
  min?: number;
  max?: number;
  align?: Align;
  color?: Color | ((row: Row, value: unknown) => Color | undefined);
  render?: (row: Row, index: number) => string;
}

export interface TableOptions<Row = Record<string, unknown>> {
  rows: Row[];
  columns: Column<Row>[];
  header?: boolean;
  headerColor?: Color;
  /** Index of the highlighted row, or -1. */
  selected?: number;
  /** First visible row; combine with `selected` for scrolling lists. */
  offset?: number;
  /**
   * Scroll so `selected` stays visible. Only the table knows how many rows fit,
   * so working the offset out here saves every caller from tracking heights.
   */
  followSelection?: boolean;
  zebra?: boolean;
  gap?: number;
  background?: Color;
  rowColor?: (row: Row, index: number) => Color | undefined;
  /** Show a scrollbar in the last column when rows overflow. */
  scrollbar?: boolean;
  onRow?: (row: Row, index: number, y: number) => void;
}

function cellValue<Row>(row: Row, column: Column<Row>, index: number): string {
  if (column.render) return column.render(row, index);
  const v = (row as Record<string, unknown>)[column.key];
  if (v === null || v === undefined) return "";
  return String(v);
}

/**
 * Where the visible window should start: the caller's offset, nudged just far
 * enough to keep the selected row on screen, and clamped to the list.
 */
export function resolveOffset(
  offset: number | undefined,
  selected: number | undefined,
  capacity: number,
  total: number,
  follow = false,
): number {
  const maxOffset = Math.max(0, total - capacity);
  let start = Math.max(0, Math.min(offset ?? 0, maxOffset));
  if (follow && selected !== undefined && capacity > 0) {
    if (selected < start) start = selected;
    else if (selected >= start + capacity) start = selected - capacity + 1;
  }
  return Math.max(0, Math.min(start, maxOffset));
}

/** A dense, column-aligned table with optional selection and scrollbar. */
export function drawTable<Row>(surface: Surface, options: TableOptions<Row>): void {
  if (surface.empty) return;
  const theme = surface.theme;
  const gap = options.gap ?? 1;
  const showHeader = options.header !== false;
  const headerRows = showHeader ? 1 : 0;
  const scrollbar = options.scrollbar ?? false;
  const bodyWidth = surface.width - (scrollbar ? 1 : 0);

  const widths = solve(
    bodyWidth,
    options.columns.map((c) => ({
      size: c.width ?? "auto",
      min: c.min ?? 1,
      max: c.max,
      intrinsic:
        c.width === undefined
          ? Math.max(
              stringWidth(c.title ?? c.key),
              ...options.rows.slice(0, 200).map((r, i) => stringWidth(cellValue(r, c, i))),
            )
          : 0,
    })),
    gap,
  );

  if (showHeader) {
    let x = 0;
    options.columns.forEach((column, i) => {
      const w = widths[i];
      if (w <= 0) return;
      surface.text(x, 0, fit(truncate(column.title ?? column.key, w), w, column.align ?? "left"), {
        fg: options.headerColor ?? theme.muted,
        bg: options.background,
        attrs: Attr.Bold,
      });
      x += w + gap;
    });
  }

  const capacity = Math.max(0, surface.height - headerRows);
  const offset = resolveOffset(options.offset, options.selected, capacity, options.rows.length, options.followSelection);
  const zebraBg = options.zebra ? elevate(theme, 0.04) : undefined;

  for (let i = 0; i < capacity; i++) {
    const rowIndex = offset + i;
    const row = options.rows[rowIndex];
    if (row === undefined) break;
    const y = i + headerRows;
    const selected = options.selected === rowIndex;
    const rowBg = selected ? theme.selection : options.zebra && rowIndex % 2 === 1 ? zebraBg : options.background;

    if (rowBg !== undefined) surface.fillRect(0, y, bodyWidth, 1, { bg: rowBg });

    let x = 0;
    options.columns.forEach((column, ci) => {
      const w = widths[ci];
      if (w <= 0) return;
      const text = cellValue(row, column, rowIndex);
      let fg: Color | undefined;
      if (selected) fg = theme.selectionText;
      else if (typeof column.color === "function") fg = column.color(row, (row as Record<string, unknown>)[column.key]);
      else if (column.color !== undefined) fg = column.color;
      else fg = options.rowColor?.(row, rowIndex);
      surface.text(x, y, fit(truncate(text, w), w, column.align ?? "left"), {
        fg: fg ?? theme.foreground,
        bg: rowBg,
        attrs: selected ? Attr.Bold : 0,
      });
      x += w + gap;
    });
    options.onRow?.(row, rowIndex, y);
  }

  if (scrollbar && options.rows.length > capacity && capacity > 0) {
    drawScrollbar(surface, surface.width - 1, headerRows, capacity, options.rows.length, offset);
  }
}

/** A one-column scrollbar. Thumb size reflects the visible fraction. */
export function drawScrollbar(
  surface: Surface,
  x: number,
  y: number,
  height: number,
  total: number,
  offset: number,
): void {
  const theme = surface.theme;
  const track = mix(theme.background, theme.border, 0.7);
  const thumbSize = Math.max(1, Math.round((height / total) * height));
  const maxOffset = Math.max(1, total - height);
  const thumbPos = Math.round((offset / maxOffset) * (height - thumbSize));
  for (let i = 0; i < height; i++) {
    const inThumb = i >= thumbPos && i < thumbPos + thumbSize;
    surface.char(x, y + i, inThumb ? "█" : "│", { fg: inThumb ? theme.accent : track });
  }
}

export interface ListOptions {
  items: (string | { label: string; color?: Color; badge?: string })[];
  selected?: number;
  offset?: number;
  /** Scroll so `selected` stays visible. */
  followSelection?: boolean;
  background?: Color;
  bullet?: string;
  scrollbar?: boolean;
}

export function drawList(surface: Surface, options: ListOptions): void {
  if (surface.empty) return;
  const theme = surface.theme;
  const width = surface.width - (options.scrollbar ? 1 : 0);
  const offset = resolveOffset(
    options.offset, options.selected, surface.height, options.items.length, options.followSelection,
  );
  for (let i = 0; i < surface.height; i++) {
    const index = offset + i;
    const raw = options.items[index];
    if (raw === undefined) break;
    const item = typeof raw === "string" ? { label: raw } : raw;
    const selected = options.selected === index;
    const bullet = options.bullet ? `${options.bullet} ` : "";
    const label = `${bullet}${item.label}`;
    if (selected) surface.fillRect(0, i, width, 1, { bg: theme.selection });
    surface.text(0, i, fit(truncate(label, width), width), {
      fg: selected ? theme.selectionText : item.color ?? theme.foreground,
      bg: selected ? theme.selection : options.background,
      attrs: selected ? Attr.Bold : 0,
    });
  }
  if (options.scrollbar && options.items.length > surface.height) {
    drawScrollbar(surface, surface.width - 1, 0, surface.height, options.items.length, offset);
  }
}

export interface TreeNode {
  label: string;
  color?: Color;
  /** Right-aligned columns, e.g. CPU% and MEM% in a process tree. */
  values?: { text: string; width: number; color?: Color; align?: Align }[];
  children?: TreeNode[];
  expanded?: boolean;
}

export interface TreeOptions {
  nodes: TreeNode[];
  selected?: number;
  offset?: number;
  /** Scroll so `selected` stays visible. */
  followSelection?: boolean;
  background?: Color;
  /** Draw the ├─ └─ connectors. */
  guides?: boolean;
  guideColor?: Color;
}

interface FlatNode {
  node: TreeNode;
  depth: number;
  last: boolean[];
}

function flatten(nodes: TreeNode[], depth: number, trail: boolean[], out: FlatNode[]): void {
  nodes.forEach((node, i) => {
    const last = i === nodes.length - 1;
    out.push({ node, depth, last: [...trail, last] });
    if (node.children && node.expanded !== false) {
      flatten(node.children, depth + 1, [...trail, last], out);
    }
  });
}

/** An indented tree with box-drawing connectors, like `pstree`. */
export function drawTree(surface: Surface, options: TreeOptions): void {
  if (surface.empty) return;
  const theme = surface.theme;
  const flat: FlatNode[] = [];
  flatten(options.nodes, 0, [], flat);

  const offset = resolveOffset(
    options.offset, options.selected, surface.height, flat.length, options.followSelection,
  );
  const guides = options.guides !== false;
  const guideColor = options.guideColor ?? mix(theme.border, theme.foreground, 0.15);

  for (let i = 0; i < surface.height; i++) {
    const index = offset + i;
    const entry = flat[index];
    if (!entry) break;
    const selected = options.selected === index;
    const bg = selected ? theme.selection : options.background;
    if (selected) surface.fillRect(0, i, surface.width, 1, { bg: theme.selection });

    let prefix = "";
    if (guides) {
      for (let d = 0; d < entry.depth; d++) prefix += entry.last[d] ? "   " : "│  ";
      prefix += entry.last[entry.depth] ? "└─ " : "├─ ";
    } else {
      prefix = "  ".repeat(entry.depth);
    }

    const valuesWidth = (entry.node.values ?? []).reduce((a, v) => a + v.width + 1, 0);
    const labelWidth = Math.max(0, surface.width - valuesWidth);
    surface.text(0, i, truncate(prefix, labelWidth), { fg: guideColor, bg });
    const px = Math.min(stringWidth(prefix), labelWidth);
    surface.text(px, i, truncate(entry.node.label, Math.max(0, labelWidth - px)), {
      fg: selected ? theme.selectionText : entry.node.color ?? theme.foreground,
      bg,
      attrs: selected ? Attr.Bold : 0,
    });

    let vx = labelWidth;
    for (const value of entry.node.values ?? []) {
      surface.text(vx, i, fit(truncate(value.text, value.width), value.width, value.align ?? "right"), {
        fg: selected ? theme.selectionText : value.color ?? theme.foreground,
        bg,
      });
      vx += value.width + 1;
    }
  }
}

export interface LogEntry {
  time?: string;
  level?: string;
  message: string;
  meta?: string;
  color?: Color;
}

export interface LogOptions {
  entries: LogEntry[];
  /** Pin to the newest entry. Default true. */
  follow?: boolean;
  offset?: number;
  background?: Color;
  levelColors?: Record<string, Color>;
  timeColor?: Color;
  metaColor?: Color;
  wrap?: boolean;
}

/** A tailing log view with colored levels. Newest at the bottom. */
export function drawLog(surface: Surface, options: LogOptions): void {
  if (surface.empty) return;
  const theme = surface.theme;
  const levels: Record<string, Color> = {
    ERROR: theme.danger,
    WARN: theme.warning,
    INFO: theme.success,
    DEBUG: theme.muted,
    TRACE: theme.muted,
    FATAL: theme.danger,
    ...options.levelColors,
  };

  const entries = options.entries;
  const follow = options.follow !== false;
  const start = follow
    ? Math.max(0, entries.length - surface.height)
    : resolveOffset(options.offset, undefined, surface.height, entries.length);

  for (let i = 0; i < surface.height; i++) {
    const entry = entries[start + i];
    if (!entry) break;
    let x = 0;
    if (entry.time) {
      x += surface.text(x, i, `${entry.time} `, { fg: options.timeColor ?? theme.muted, bg: options.background });
    }
    if (entry.level) {
      const color = levels[entry.level.toUpperCase()] ?? theme.foreground;
      x += surface.text(x, i, fit(entry.level.toUpperCase(), 5), { fg: color, bg: options.background, attrs: Attr.Bold });
      x += surface.text(x, i, " ", { bg: options.background });
    }
    const metaWidth = entry.meta ? stringWidth(entry.meta) + 1 : 0;
    const msgWidth = Math.max(0, surface.width - x - metaWidth);
    surface.text(x, i, truncate(entry.message, msgWidth), {
      fg: entry.color ?? theme.foreground,
      bg: options.background,
    });
    if (entry.meta && metaWidth < surface.width) {
      surface.text(surface.width - metaWidth + 1, i, entry.meta, {
        fg: options.metaColor ?? theme.muted,
        bg: options.background,
      });
    }
  }
}

File: ports/python/hqtui/widgets/table.py
"""Dense data widgets: tables, lists, trees, tailing logs and the scrollbar
they share.

The reference implementation is generic over a row type and reads cells with a
key or a ``render`` callback. Python takes the cells already stringified —
``TableRow(("1", "systemd", "0.1"))`` — which keeps the widget out of the
business of formatting someone else's objects.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Mapping, Sequence

from ..buffer import Attrs, Style
from ..color import Color, round_half_up
from ..layout import Constraint, solve
from ..surface import Surface, TextOptions
from ..theme import elevate
from ..unicode import Align, fit, string_width, truncate

__all__ = [
    "ListItem",
    "ListOptions",
    "LogEntry",
    "LogOptions",
    "TableColumn",
    "TableOptions",
    "TableRow",
    "TreeNode",
    "TreeOptions",
    "TreeValue",
    "draw_list",
    "draw_log",
    "draw_scrollbar",
    "draw_table",
    "draw_tree",
    "resolve_offset",
]


@dataclass(frozen=True, slots=True)
class TableColumn:
    title: str = ""
    width: "int | str | None" = None
    min: int | None = None
    max: int | None = None
    align: "Align | str" = Align.LEFT
    color: Color | None = None


@dataclass(frozen=True, slots=True)
class TableRow:
    cells: Sequence[str] = ()
    color: Color | None = None
    """Colors the whole row, unless the column or the cell says otherwise."""
    cell_colors: Sequence[Color | None] = ()
    """Per-cell colors, standing in for the reference's per-column color
    function. Where an entry exists for a column it wins outright, including
    when it is ``None`` — which is what that callback returning ``undefined``
    does there."""


@dataclass(frozen=True, slots=True)
class TableOptions:
    rows: Sequence[TableRow] = ()
    columns: Sequence[TableColumn] = ()
    header: bool = True
    header_color: Color | None = None
    selected: int | None = None
    """Index of the highlighted row."""
    offset: int | None = None
    """First visible row; combine with ``selected`` for scrolling lists."""
    follow_selection: bool = False
    """Scroll so ``selected`` stays visible. Only the table knows how many rows
    fit, so working the offset out here saves every caller from tracking
    heights."""
    zebra: bool = False
    gap: int = 1
    background: Color | None = None
    scrollbar: bool = False
    """Show a scrollbar in the last column when rows overflow."""


def resolve_offset(
    offset: int | None, selected: int | None, capacity: int, total: int, follow: bool = False
) -> int:
    """Where the visible window should start: the caller's offset, nudged just
    far enough to keep the selected row on screen, and clamped to the list."""
    max_offset = max(0, total - capacity)
    start = max(0, min(offset or 0, max_offset))
    if follow and selected is not None and capacity > 0:
        if selected < start:
            start = selected
        elif selected >= start + capacity:
            start = selected - capacity + 1
    return max(0, min(start, max_offset))


def draw_table(surface: Surface, options: TableOptions) -> None:
    """A dense, column-aligned table with optional selection and scrollbar."""
    if surface.empty:
        return
    theme = surface.theme
    gap = options.gap
    header_rows = 1 if options.header else 0
    body_width = surface.width - (1 if options.scrollbar else 0)

    constraints: list[Constraint] = []
    for ci, c in enumerate(options.columns):
        intrinsic = 0
        if c.width is None:
            # The reference samples the first 200 rows, which is what keeps an
            # auto column cheap on a long table.
            widest = max(
                (
                    string_width(r.cells[ci]) if ci < len(r.cells) else 0
                    for r in options.rows[:200]
                ),
                default=0,
            )
            intrinsic = max(string_width(c.title), widest)
        constraints.append(
            Constraint(
                size=c.width if c.width is not None else "auto",
                min=c.min if c.min is not None else 1,
                max=c.max,
                intrinsic=intrinsic,
            )
        )
    widths = solve(body_width, constraints, gap)

    if options.header:
        header_color = (
            options.header_color if options.header_color is not None else theme.muted
        )
        x = 0
        for i, column in enumerate(options.columns):
            w = widths[i]
            if w <= 0:
                continue
            surface.text(
                x, 0, fit(truncate(column.title, w), w, column.align),
                TextOptions(fg=header_color, bg=options.background, attrs=Attrs.BOLD),
            )
            x += w + gap

    capacity = max(0, surface.height - header_rows)
    offset = resolve_offset(
        options.offset, options.selected, capacity, len(options.rows), options.follow_selection
    )
    zebra_bg = elevate(theme, 0.04) if options.zebra else None

    for i in range(capacity):
        row_index = offset + i
        if row_index >= len(options.rows):
            break
        row = options.rows[row_index]
        y = i + header_rows
        selected = options.selected == row_index

        if selected:
            row_bg = theme.selection
        elif options.zebra and row_index % 2 == 1:
            row_bg = zebra_bg
        else:
            row_bg = options.background
        if row_bg is not None:
            surface.fill_rect(0, y, body_width, 1, Style(bg=row_bg))

        x = 0
        for ci, column in enumerate(options.columns):
            w = widths[ci]
            if w <= 0:
                continue
            text = row.cells[ci] if ci < len(row.cells) else ""
            if selected:
                fg = theme.selection_text
            elif ci < len(row.cell_colors):
                fg = row.cell_colors[ci]
            elif column.color is not None:
                fg = column.color
            else:
                fg = row.color
            surface.text(
                x, y, fit(truncate(text, w), w, column.align),
                TextOptions(
                    fg=fg if fg is not None else theme.foreground,
                    bg=row_bg,
                    attrs=Attrs.BOLD if selected else 0,
                ),
            )
            x += w + gap

    if options.scrollbar and len(options.rows) > capacity and capacity > 0:
        draw_scrollbar(
            surface, surface.width - 1, header_rows, capacity, len(options.rows), offset
        )


def draw_scrollbar(
    surface: Surface, x: int, y: int, height: int, total: int, offset: int
) -> None:
    """A one-column scrollbar. Thumb size reflects the visible fraction."""
    theme = surface.theme
    track = theme.background.mix(theme.border, 0.7)
    thumb_size = max(1, int(round_half_up(height / total * height)))
    max_offset = max(1, total - height)
    thumb_pos = int(round_half_up(offset / max_offset * (height - thumb_size)))
    for i in range(height):
        in_thumb = thumb_pos <= i < thumb_pos + thumb_size
        surface.char(
            x, y + i, "█" if in_thumb else "│",
            Style(fg=theme.accent if in_thumb else track),
        )


@dataclass(frozen=True, slots=True)
class ListItem:
    label: str = ""
    color: Color | None = None
    badge: str = ""


@dataclass(frozen=True, slots=True)
class ListOptions:
    items: Sequence["ListItem | str"] = ()
    selected: int | None = None
    offset: int | None = None
    follow_selection: bool = False
    background: Color | None = None
    bullet: str = ""
    scrollbar: bool = False


def draw_list(surface: Surface, options: ListOptions) -> None:
    if surface.empty:
        return
    theme = surface.theme
    width = surface.width - (1 if options.scrollbar else 0)
    offset = resolve_offset(
        options.offset, options.selected, surface.height, len(options.items),
        options.follow_selection,
    )
    for i in range(surface.height):
        index = offset + i
        if index >= len(options.items):
            break
        raw = options.items[index]
        item = ListItem(label=raw) if isinstance(raw, str) else raw
        selected = options.selected == index
        bullet = f"{options.bullet} " if options.bullet else ""
        if selected:
            surface.fill_rect(0, i, width, 1, Style(bg=theme.selection))
        surface.text(
            0, i, fit(truncate(bullet + item.label, width), width, Align.LEFT),
            TextOptions(
                fg=theme.selection_text if selected else (item.color or theme.foreground),
                bg=theme.selection if selected else options.background,
                attrs=Attrs.BOLD if selected else 0,
            ),
        )
    if options.scrollbar and len(options.items) > surface.height:
        draw_scrollbar(
            surface, surface.width - 1, 0, surface.height, len(options.items), offset
        )


@dataclass(frozen=True, slots=True)
class TreeValue:
    text: str = ""
    width: int = 0
    color: Color | None = None
    align: "Align | str" = Align.RIGHT


@dataclass(frozen=True, slots=True)
class TreeNode:
    label: str = ""
    color: Color | None = None
    values: Sequence[TreeValue] = ()
    """Right-aligned columns, e.g. CPU% and MEM% in a process tree."""
    children: Sequence["TreeNode"] = ()
    expanded: bool = True


@dataclass(frozen=True, slots=True)
class TreeOptions:
    nodes: Sequence[TreeNode] = ()
    selected: int | None = None
    offset: int | None = None
    follow_selection: bool = False
    background: Color | None = None
    guides: bool = True
    """Draw the ├─ └─ connectors."""
    guide_color: Color | None = None


def _flatten(
    nodes: Sequence[TreeNode], depth: int, trail: list[bool], out: list
) -> None:
    for i, node in enumerate(nodes):
        last = i == len(nodes) - 1
        here = [*trail, last]
        out.append((node, depth, here))
        if node.children and node.expanded:
            _flatten(node.children, depth + 1, here, out)


def draw_tree(surface: Surface, options: TreeOptions) -> None:
    """An indented tree with box-drawing connectors, like ``pstree``."""
    if surface.empty:
        return
    theme = surface.theme
    flat: list = []
    _flatten(options.nodes, 0, [], flat)

    offset = resolve_offset(
        options.offset, options.selected, surface.height, len(flat), options.follow_selection
    )
    guide_color = (
        options.guide_color
        if options.guide_color is not None
        else theme.border.mix(theme.foreground, 0.15)
    )

    for i in range(surface.height):
        index = offset + i
        if index >= len(flat):
            break
        node, depth, last = flat[index]
        selected = options.selected == index
        bg = theme.selection if selected else options.background
        if selected:
            surface.fill_rect(0, i, surface.width, 1, Style(bg=theme.selection))

        if options.guides:
            prefix = "".join("   " if last[d] else "│  " for d in range(depth))
            prefix += "└─ " if last[depth] else "├─ "
        else:
            prefix = "  " * depth

        values_width = sum(v.width + 1 for v in node.values)
        label_width = max(0, surface.width - values_width)
        surface.text(
            0, i, truncate(prefix, label_width), TextOptions(fg=guide_color, bg=bg)
        )
        px = min(string_width(prefix), label_width)
        surface.text(
            px, i, truncate(node.label, max(0, label_width - px)),
            TextOptions(
                fg=theme.selection_text if selected else (node.color or theme.foreground),
                bg=bg,
                attrs=Attrs.BOLD if selected else 0,
            ),
        )

        vx = label_width
        for value in node.values:
            surface.text(
                vx, i, fit(truncate(value.text, value.width), value.width, value.align),
                TextOptions(
                    fg=theme.selection_text if selected else (value.color or theme.foreground),
                    bg=bg,
                ),
            )
            vx += value.width + 1


@dataclass(frozen=True, slots=True)
class LogEntry:
    message: str = ""
    time: str = ""
    level: str = ""
    meta: str = ""
    color: Color | None = None


@dataclass(frozen=True, slots=True)
class LogOptions:
    entries: Sequence[LogEntry] = ()
    follow: bool = True
    """Pin to the newest entry. Set False to scroll with ``offset``."""
    offset: int | None = None
    from_end: int = 0
    """Lines to scroll back from the newest entry. The natural control for a
    tailing log: only the widget knows how many rows fit, so an absolute offset
    makes small scrolls near the bottom clamp to nothing."""
    scrollbar: bool = False
    background: Color | None = None
    level_colors: Mapping[str, Color] = field(default_factory=dict)
    time_color: Color | None = None
    meta_color: Color | None = None


def draw_log(surface: Surface, options: LogOptions) -> None:
    """A tailing log view with colored levels. Newest at the bottom."""
    if surface.empty:
        return
    theme = surface.theme
    defaults = {
        "ERROR": theme.danger,
        "WARN": theme.warning,
        "INFO": theme.success,
        "DEBUG": theme.muted,
        "TRACE": theme.muted,
        "FATAL": theme.danger,
    }
    levels = {**defaults, **options.level_colors}

    entries = options.entries
    width = surface.width - (1 if options.scrollbar else 0)
    from_end = max(0, options.from_end)
    max_start = max(0, len(entries) - surface.height)
    if options.follow or from_end > 0:
        start = max(0, min(max_start, len(entries) - surface.height - from_end))
    else:
        start = resolve_offset(options.offset, None, surface.height, len(entries))

    for i in range(surface.height):
        if start + i >= len(entries):
            break
        entry = entries[start + i]
        x = 0
        if entry.time:
            x += surface.text(
                x, i, f"{entry.time} ",
                TextOptions(
                    fg=options.time_color if options.time_color is not None else theme.muted,
                    bg=options.background,
                ),
            )
        if entry.level:
            color = levels.get(entry.level.upper(), theme.foreground)
            x += surface.text(
                x, i, fit(entry.level.upper(), 5, Align.LEFT),
                TextOptions(fg=color, bg=options.background, attrs=Attrs.BOLD),
            )
            x += surface.text(x, i, " ", TextOptions(bg=options.background))
        meta_width = string_width(entry.meta) + 1 if entry.meta else 0
        msg_width = max(0, width - x - meta_width)
        surface.text(
            x, i, truncate(entry.message, msg_width),
            TextOptions(
                fg=entry.color if entry.color is not None else theme.foreground,
                bg=options.background,
            ),
        )
        if entry.meta and meta_width < width:
            surface.text(
                width - meta_width + 1, i, entry.meta,
                TextOptions(
                    fg=options.meta_color if options.meta_color is not None else theme.muted,
                    bg=options.background,
                ),
            )
